diff --git a/quickstep/dagger/com/android/launcher3/dagger/AppActivityContextModule.kt b/quickstep/dagger/com/android/launcher3/dagger/AppActivityContextModule.kt
index a703b25a29..105f31b5e5 100644
--- a/quickstep/dagger/com/android/launcher3/dagger/AppActivityContextModule.kt
+++ b/quickstep/dagger/com/android/launcher3/dagger/AppActivityContextModule.kt
@@ -16,10 +16,18 @@
 
 package com.android.launcher3.dagger
 
+import com.android.launcher3.taskbar.allapps.TaskbarSearchSessionController
+import com.android.launcher3.taskbar.allapps.TaskbarSearchSessionControllerImpl
+import dagger.Binds
 import dagger.Module
 
 /**
  * Module containing the app specific [ActivityContext] bindings for the final derivate app, an
  * implementation of this module should be included in the final app code.
  */
-@Module abstract class AppActivityContextModule {}
+@Module abstract class AppActivityContextModule {
+    @Binds
+    abstract fun bindTaskbarSearchSessionController(
+        impl: TaskbarSearchSessionControllerImpl
+    ): TaskbarSearchSessionController
+}
diff --git a/quickstep/src/com/android/launcher3/model/QuickstepModelDelegate.java b/quickstep/src/com/android/launcher3/model/QuickstepModelDelegate.java
index 3bbd5cb9fb..c73f7c23c1 100644
--- a/quickstep/src/com/android/launcher3/model/QuickstepModelDelegate.java
+++ b/quickstep/src/com/android/launcher3/model/QuickstepModelDelegate.java
@@ -224,8 +224,8 @@ public class QuickstepModelDelegate extends ModelDelegate {
                         "Item info: %s found with invalid container: %s",
                         info,
                         containerInfo));
+                return null;
             }
-            // Allow crash to help debug b/173838775
             return (CollectionInfo) containerInfo;
         }
         return null;
diff --git a/quickstep/src/com/android/launcher3/taskbar/allapps/TaskbarSearchSessionControllerImpl.kt b/quickstep/src/com/android/launcher3/taskbar/allapps/TaskbarSearchSessionControllerImpl.kt
new file mode 100644
index 0000000000..ebbf2a45de
--- /dev/null
+++ b/quickstep/src/com/android/launcher3/taskbar/allapps/TaskbarSearchSessionControllerImpl.kt
@@ -0,0 +1,19 @@
+package com.android.launcher3.taskbar.allapps
+
+import com.android.launcher3.allapps.search.SearchSessionManager
+import com.android.launcher3.dagger.ActivityContextSingleton
+import com.android.launcher3.views.ActivityContext
+import javax.inject.Inject
+
+@ActivityContextSingleton
+class TaskbarSearchSessionControllerImpl @Inject constructor(
+    val mActivityContext: ActivityContext
+) : TaskbarSearchSessionController() {
+    override fun canHandleBackInvoked(): Boolean {
+        return SearchSessionManager.handleAllAppsSearchBackInvoked(mActivityContext, false)
+    }
+
+    override fun handleBackInvoked(): Boolean {
+        return SearchSessionManager.handleAllAppsSearchBackInvoked(mActivityContext, true)
+    }
+}
diff --git a/quickstep/src/com/android/launcher3/uioverrides/QuickstepLauncher.java b/quickstep/src/com/android/launcher3/uioverrides/QuickstepLauncher.java
index 51d24c44ea..cd16202824 100644
--- a/quickstep/src/com/android/launcher3/uioverrides/QuickstepLauncher.java
+++ b/quickstep/src/com/android/launcher3/uioverrides/QuickstepLauncher.java
@@ -134,6 +134,7 @@ import com.android.launcher3.Utilities;
 import com.android.launcher3.Workspace;
 import com.android.launcher3.accessibility.LauncherAccessibilityDelegate;
 import com.android.launcher3.allapps.AllAppsRecyclerView;
+import com.android.launcher3.allapps.search.SearchSessionManager;
 import com.android.launcher3.anim.PendingAnimation;
 import com.android.launcher3.apppairs.AppPairIcon;
 import com.android.launcher3.appprediction.PredictionRowView;
@@ -383,6 +384,13 @@ public class QuickstepLauncher extends Launcher implements RecentsViewContainer,
         mBubbleFeatureConfig = new BubbleFeatureConfigImpl(this, getDesktopState(this));
     }
 
+    @Override
+    protected void onStateBack() {
+        if (!SearchSessionManager.handleAllAppsSearchBackInvoked(this, true)) {
+            super.onStateBack();
+        }
+    }
+
     @Override
     public void logAppLaunch(StatsLogManager statsLogManager, ItemInfo info,
             InstanceId instanceId) {
diff --git a/quickstep/src/com/android/quickstep/AllAppsActionManager.kt b/quickstep/src/com/android/quickstep/AllAppsActionManager.kt
index 6b60e4d48a..95eedaae96 100644
--- a/quickstep/src/com/android/quickstep/AllAppsActionManager.kt
+++ b/quickstep/src/com/android/quickstep/AllAppsActionManager.kt
@@ -23,6 +23,7 @@ import android.content.Context
 import android.graphics.drawable.Icon
 import android.provider.Settings
 import android.provider.Settings.Secure.USER_SETUP_COMPLETE
+import android.util.Log
 import android.view.accessibility.AccessibilityManager
 import com.android.launcher3.R
 import com.android.launcher3.concurrent.annotations.LightweightBackground
@@ -58,6 +59,8 @@ constructor(
 
     private var onSettingsChangeSafeCloseable: SafeCloseable? = null
 
+    private var isDestroyed = false
+
     init {
         onSettingsChangeSafeCloseable =
             SettingsCache.INSTANCE[context].getListenableRef(USER_SETUP_COMPLETE_URI).forEach(
@@ -114,6 +117,11 @@ constructor(
 
     private fun updateSystemAction() {
         synchronized(this) {
+            if (isDestroyed) {
+                Log.w(TAG, "updateSystemAction called when destroyed", Throwable())
+                return
+            }
+
             val isInSetupFlow = isSetupUiVisible || !isUserSetupComplete
             val shouldRegisterAction =
                 (isHomeAndOverviewSame || isTaskbarPresent) && !isInSetupFlow && isUserUnlocked
@@ -121,27 +129,36 @@ constructor(
             isActionRegistered = shouldRegisterAction
 
             bgExecutor.execute {
-                val accessibilityManager =
-                    context.getSystemService(AccessibilityManager::class.java) ?: return@execute
-                if (shouldRegisterAction) {
-                    val allAppsPendingIntent = PendingIntent(allAppsIntentSenderProvider.get())
-                    accessibilityManager.registerSystemAction(
-                        RemoteAction(
-                            Icon.createWithResource(context, R.drawable.ic_apps),
-                            context.getString(R.string.all_apps_label),
-                            context.getString(R.string.all_apps_label),
-                            allAppsPendingIntent,
-                        ),
-                        GLOBAL_ACTION_ACCESSIBILITY_ALL_APPS,
-                    )
-                    quickstepKeyGestureEventsManager.registerAllAppsKeyGestureEvent(
-                        allAppsPendingIntent
-                    )
-                } else {
-                    accessibilityManager.unregisterSystemAction(
-                        GLOBAL_ACTION_ACCESSIBILITY_ALL_APPS
-                    )
-                    quickstepKeyGestureEventsManager.unregisterAllAppsKeyGestureEvent()
+                synchronized(this@AllAppsActionManager) {
+                    if (isDestroyed) {
+                        isActionRegistered = false
+                        Log.w(TAG, "updateSystemAction bgExecutor task running " +
+                                "when destroyed; shouldRegisterAction $shouldRegisterAction")
+                        return@execute
+                    }
+
+                    val accessibilityManager =
+                        context.getSystemService(AccessibilityManager::class.java) ?: return@execute
+                    if (shouldRegisterAction) {
+                        val allAppsPendingIntent = PendingIntent(allAppsIntentSenderProvider.get())
+                        accessibilityManager.registerSystemAction(
+                            RemoteAction(
+                                Icon.createWithResource(context, R.drawable.ic_apps),
+                                context.getString(R.string.all_apps_label),
+                                context.getString(R.string.all_apps_label),
+                                allAppsPendingIntent,
+                            ),
+                            GLOBAL_ACTION_ACCESSIBILITY_ALL_APPS,
+                        )
+                        quickstepKeyGestureEventsManager.registerAllAppsKeyGestureEvent(
+                            allAppsPendingIntent
+                        )
+                    } else {
+                        accessibilityManager.unregisterSystemAction(
+                            GLOBAL_ACTION_ACCESSIBILITY_ALL_APPS
+                        )
+                        quickstepKeyGestureEventsManager.unregisterAllAppsKeyGestureEvent()
+                    }
                 }
             }
         }
@@ -149,6 +166,7 @@ constructor(
 
     fun onDestroy() {
         synchronized(this) {
+            isDestroyed = true
             isActionRegistered = false
             context
                 .getSystemService(AccessibilityManager::class.java)
@@ -168,4 +186,8 @@ constructor(
         pw.println("\tisActionRegistered=$isActionRegistered")
         pw.println("\tisUserUnlocked=$isUserUnlocked")
     }
+
+    companion object {
+        private const val TAG = "AllAppsActionManager"
+    }
 }
diff --git a/res/values/dimens.xml b/res/values/dimens.xml
index 3f631f6116..cb7d6b3a44 100644
--- a/res/values/dimens.xml
+++ b/res/values/dimens.xml
@@ -112,8 +112,8 @@
 
     <!-- All Apps -->
     <dimen name="all_apps_search_bar_field_height">48dp</dimen>
-    <!-- all_apps_search_bar_field_height / 2 -->
-    <dimen name="all_apps_search_bar_content_overlap">24dp</dimen>
+    <!-- Revert commit to restore to 24dp if a prediction service exists -->
+    <dimen name="all_apps_search_bar_content_overlap">0dp</dimen>
     <dimen name="all_apps_search_bar_bottom_padding">30dp</dimen>
     <dimen name="all_apps_empty_search_message_top_offset">40dp</dimen>
     <dimen name="all_apps_search_top_row_extra_height">4dp</dimen>
diff --git a/res/values/strings.xml b/res/values/strings.xml
index f709a698e6..6e9f231cc4 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -21,7 +21,7 @@
     <skip />
 
     <!-- Application name -->
-    <string name="app_name">Launcher3</string>
+    <string name="app_name">AIOS Home</string>
     <!-- Work folder name -->
     <string name="work_folder_name">Work</string>
     <!-- Displayed when user selects a shortcut for an app that was uninstalled [CHAR_LIMIT=none]-->
diff --git a/res/xml/default_workspace_2x2.xml b/res/xml/default_workspace_2x2.xml
new file mode 100644
index 0000000000..51d9e07005
--- /dev/null
+++ b/res/xml/default_workspace_2x2.xml
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2015 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+
+<favorites xmlns:launcher="http://schemas.android.com/apk/res-auto/com.android.launcher3">
+
+    <!-- Hotseat (We use the screen as the position of the item in the hotseat) -->
+    <!-- Messaging, Dialer -->
+
+    <resolve
+        launcher:container="-101"
+        launcher:screen="0"
+        launcher:x="0"
+        launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_MESSAGING;end" />
+        <favorite launcher:uri="sms:" />
+        <favorite launcher:uri="smsto:" />
+        <favorite launcher:uri="mms:" />
+        <favorite launcher:uri="mmsto:" />
+    </resolve>
+
+    <resolve
+        launcher:container="-101"
+        launcher:screen="2"
+        launcher:x="2"
+        launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.DIAL;end" />
+        <favorite launcher:uri="tel:123" />
+        <favorite launcher:uri="#Intent;action=android.intent.action.CALL_BUTTON;end" />
+    </resolve>
+
+    <!-- Bottom row -->
+    <resolve
+        launcher:screen="0"
+        launcher:x="0"
+        launcher:y="-1" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_EMAIL;end" />
+        <favorite launcher:uri="mailto:" />
+    </resolve>
+
+    <resolve
+        launcher:screen="0"
+        launcher:x="1"
+        launcher:y="-1" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_GALLERY;end" />
+        <favorite launcher:uri="#Intent;type=images/*;end" />
+    </resolve>
+
+</favorites>
diff --git a/res/xml/default_workspace_4x5.xml b/res/xml/default_workspace_4x5.xml
new file mode 100644
index 0000000000..bf3c62c77d
--- /dev/null
+++ b/res/xml/default_workspace_4x5.xml
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2009 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+
+<favorites xmlns:launcher="http://schemas.android.com/apk/res-auto/com.android.launcher3">
+
+    <!-- Hotseat (We use the screen as the position of the item in the hotseat) -->
+    <!-- Dialer, Messaging, Browser, Camera -->
+    <resolve
+        launcher:container="-101"
+        launcher:screen="0"
+        launcher:x="0"
+        launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.DIAL;end" />
+        <favorite launcher:uri="tel:123" />
+        <favorite launcher:uri="#Intent;action=android.intent.action.CALL_BUTTON;end" />
+    </resolve>
+
+    <resolve
+        launcher:container="-101"
+        launcher:screen="1"
+        launcher:x="1"
+        launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_MESSAGING;end" />
+        <favorite launcher:uri="sms:" />
+        <favorite launcher:uri="smsto:" />
+        <favorite launcher:uri="mms:" />
+        <favorite launcher:uri="mmsto:" />
+    </resolve>
+
+    <resolve
+        launcher:container="-101"
+        launcher:screen="2"
+        launcher:x="2"
+        launcher:y="0" >
+        <favorite
+            launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_BROWSER;end" />
+        <favorite launcher:uri="http://www.example.com/" />
+    </resolve>
+
+    <resolve
+        launcher:container="-101"
+        launcher:screen="3"
+        launcher:x="3"
+        launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.media.action.STILL_IMAGE_CAMERA;end" />
+        <favorite launcher:uri="#Intent;action=android.intent.action.CAMERA_BUTTON;end" />
+    </resolve>
+
+    <!-- Bottom row -->
+    <resolve
+        launcher:screen="0"
+        launcher:x="0"
+        launcher:y="-1" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_EMAIL;end" />
+        <favorite launcher:uri="mailto:" />
+    </resolve>
+
+    <resolve
+        launcher:screen="0"
+        launcher:x="1"
+        launcher:y="-1" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_GALLERY;end" />
+        <favorite launcher:uri="#Intent;type=images/*;end" />
+    </resolve>
+
+    <resolve
+        launcher:screen="0"
+        launcher:x="3"
+        launcher:y="-1" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_MARKET;end" />
+        <favorite launcher:uri="market://details?id=com.android.launcher" />
+    </resolve>
+
+</favorites>
diff --git a/res/xml/default_workspace_5x5.xml b/res/xml/default_workspace_5x5.xml
index b4ac8f63df..b260524b26 100644
--- a/res/xml/default_workspace_5x5.xml
+++ b/res/xml/default_workspace_5x5.xml
@@ -45,6 +45,7 @@
         launcher:screen="2"
         launcher:x="2"
         launcher:y="0" >
+        <favorite launcher:uri="#Intent;action=android.intent.action.ASSIST;end" />
         <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_MAPS;end" />
         <favorite launcher:uri="#Intent;action=android.intent.action.MAIN;category=android.intent.category.APP_MUSIC;end" />
     </resolve>
diff --git a/res/xml/device_profiles.xml b/res/xml/device_profiles.xml
index 934c93c645..55b29ad5ed 100644
--- a/res/xml/device_profiles.xml
+++ b/res/xml/device_profiles.xml
@@ -17,6 +17,39 @@
 
 <profiles xmlns:launcher="http://schemas.android.com/apk/res-auto" >
 
+    <grid-option
+        launcher:name="2_by_2"
+        launcher:numRows="2"
+        launcher:numColumns="2"
+        launcher:numFolderRows="2"
+        launcher:numFolderColumns="2"
+        launcher:numHotseatIcons="2"
+        launcher:dbFile="launcher_2_by_2.db"
+        launcher:defaultLayoutId="@xml/default_workspace_2x2"
+        launcher:deviceCategory="phone" >
+
+        <display-option
+            launcher:name="Super Short Stubby"
+            launcher:minWidthDps="200"
+            launcher:minHeightDps="200"
+            launcher:iconImageSize="48"
+            launcher:iconTextSize="13.0"
+            launcher:allAppsBorderSpace="16"
+            launcher:allAppsCellHeight="104"
+            launcher:canBeDefault="true" />
+
+        <display-option
+            launcher:name="Shorter Stubby"
+            launcher:minWidthDps="200"
+            launcher:minHeightDps="300"
+            launcher:iconImageSize="48"
+            launcher:iconTextSize="13.0"
+            launcher:allAppsBorderSpace="16"
+            launcher:allAppsCellHeight="104"
+            launcher:canBeDefault="true" />
+
+    </grid-option>
+
     <grid-option
         launcher:name="3_by_3"
         launcher:numRows="3"
@@ -115,6 +144,28 @@
 
     </grid-option>
 
+    <grid-option
+        launcher:name="4_by_5"
+        launcher:numRows="5"
+        launcher:numColumns="4"
+        launcher:numFolderRows="3"
+        launcher:numFolderColumns="3"
+        launcher:numHotseatIcons="4"
+        launcher:dbFile="launcher_4_by_5.db"
+        launcher:defaultLayoutId="@xml/default_workspace_4x5"
+        launcher:deviceCategory="phone|multi_display" >
+
+        <display-option
+            launcher:name="Pixel 2021"
+            launcher:minWidthDps="367"
+            launcher:minHeightDps="838"
+            launcher:allAppsBorderSpace="16"
+            launcher:allAppsCellHeight="104"
+            launcher:iconImageSize="66"
+            launcher:iconTextSize="14.0"
+            launcher:canBeDefault="true" />
+    </grid-option>
+
     <grid-option
         launcher:name="5_by_5"
         launcher:numRows="5"
diff --git a/src/com/android/launcher3/InvariantDeviceProfile.java b/src/com/android/launcher3/InvariantDeviceProfile.java
index 5c3bfb6e80..3a748767b3 100644
--- a/src/com/android/launcher3/InvariantDeviceProfile.java
+++ b/src/com/android/launcher3/InvariantDeviceProfile.java
@@ -302,9 +302,11 @@ public class InvariantDeviceProfile {
                 onConfigChanged();
             }
         };
-        prefs.addListener(prefListener, FIXED_LANDSCAPE_MODE, ENABLE_TWOLINE_ALLAPPS_TOGGLE);
+        prefs.addListener(prefListener, FIXED_LANDSCAPE_MODE, ENABLE_TWOLINE_ALLAPPS_TOGGLE,
+                WORKSPACE_ITEMS_LABEL_HIDDEN);
         lifeCycle.addCloseable(() -> prefs.removeListener(prefListener,
-                FIXED_LANDSCAPE_MODE, ENABLE_TWOLINE_ALLAPPS_TOGGLE));
+                FIXED_LANDSCAPE_MODE, ENABLE_TWOLINE_ALLAPPS_TOGGLE,
+                WORKSPACE_ITEMS_LABEL_HIDDEN));
 
         SimpleBroadcastReceiver localeReceiver = new SimpleBroadcastReceiver(context,
                 mMainExecutor, i -> onConfigChanged());
diff --git a/src/com/android/launcher3/allapps/ActivityAllAppsContainerView.java b/src/com/android/launcher3/allapps/ActivityAllAppsContainerView.java
index 49ec0f63e1..b80041338c 100644
--- a/src/com/android/launcher3/allapps/ActivityAllAppsContainerView.java
+++ b/src/com/android/launcher3/allapps/ActivityAllAppsContainerView.java
@@ -79,6 +79,7 @@ import com.android.launcher3.Utilities;
 import com.android.launcher3.allapps.BaseAllAppsAdapter.AdapterItem;
 import com.android.launcher3.allapps.search.AllAppsSearchUiDelegate;
 import com.android.launcher3.allapps.search.SearchAdapterProvider;
+import com.android.launcher3.allapps.search.SearchSessionManager;
 import com.android.launcher3.config.FeatureFlags;
 import com.android.launcher3.keyboard.FocusedItemDecorator;
 import com.android.launcher3.keyboard.ViewGroupFocusHelper;
@@ -365,7 +366,7 @@ public class ActivityAllAppsContainerView<T extends Context & ActivityContext>
     /**
      * Sets results list for search
      */
-    public void setSearchResults(ArrayList<AdapterItem> results) {
+    public void setSearchResults(List<AdapterItem> results) {
         getMainAdapterProvider().clearHighlightedItem();
         if (getSearchResultList().setSearchResults(results)) {
             getSearchRecyclerView().onSearchResultsChanged();
@@ -560,7 +561,8 @@ public class ActivityAllAppsContainerView<T extends Context & ActivityContext>
      * @return {@code true} if back gesture should exit search rather than change launcher state.
       */
     public boolean shouldBackExitSearch() {
-        return isSearching();
+        return isSearching() && SearchSessionManager.handleAllAppsSearchBackInvoked(
+                mActivityContext, false);
     }
 
     @Override
diff --git a/src/com/android/launcher3/allapps/AlphabeticalAppsList.java b/src/com/android/launcher3/allapps/AlphabeticalAppsList.java
index 1bdd53d5bb..33dc0cf50f 100644
--- a/src/com/android/launcher3/allapps/AlphabeticalAppsList.java
+++ b/src/com/android/launcher3/allapps/AlphabeticalAppsList.java
@@ -217,7 +217,7 @@ public class AlphabeticalAppsList implements AllAppsStore.OnUpdateListener {
     /**
      * Sets results list for search
      */
-    public boolean setSearchResults(ArrayList<AdapterItem> results) {
+    public boolean setSearchResults(List<AdapterItem> results) {
         if (Objects.equals(results, mSearchResults)) {
             return false;
         }
diff --git a/src/com/android/launcher3/allapps/SearchUiManager.java b/src/com/android/launcher3/allapps/SearchUiManager.java
index 4dd820e10d..36f8ce4a20 100644
--- a/src/com/android/launcher3/allapps/SearchUiManager.java
+++ b/src/com/android/launcher3/allapps/SearchUiManager.java
@@ -25,6 +25,11 @@ import com.android.launcher3.ExtendedEditText;
  * Interface for controlling the Apps search UI.
  */
 public interface SearchUiManager {
+    default boolean shouldInterceptBackButton() {
+        // In NexusLauncher, this would be stored and accessed from OneSearchSessionManager.
+        // Doing it here makes it simpler, i.e. avoids us needing to manage a SessionManagerObject.
+        return false;
+    }
 
     /**
      * Initializes the search manager.
diff --git a/src/com/android/launcher3/allapps/search/AllAppsSearchBarController.java b/src/com/android/launcher3/allapps/search/AllAppsSearchBarController.java
index fe11ee20d1..4b05c08bbf 100644
--- a/src/com/android/launcher3/allapps/search/AllAppsSearchBarController.java
+++ b/src/com/android/launcher3/allapps/search/AllAppsSearchBarController.java
@@ -134,6 +134,10 @@ public class AllAppsSearchBarController
 
     @Override
     public boolean onBackKey() {
+        // Note: This doesn't seem to be called from ExtendedEditText#onkeyPreIme, likely because
+        // predictive back is being used.
+        Log.d(TAG, "onBackKey");
+
         // Only hide the search field if there is no query
         String query = Utilities.trim(mInput.getEditableText().toString());
         if (query.isEmpty()) {
diff --git a/src/com/android/launcher3/allapps/search/AppsSearchContainerLayout.java b/src/com/android/launcher3/allapps/search/AppsSearchContainerLayout.java
index 45b4071ec9..f0785115d7 100644
--- a/src/com/android/launcher3/allapps/search/AppsSearchContainerLayout.java
+++ b/src/com/android/launcher3/allapps/search/AppsSearchContainerLayout.java
@@ -24,13 +24,16 @@ import static com.android.launcher3.icons.IconNormalizer.ICON_VISIBLE_AREA_FACTO
 
 import android.content.Context;
 import android.graphics.Rect;
+import android.text.Editable;
 import android.text.Selection;
 import android.text.SpannableStringBuilder;
+import android.text.TextWatcher;
 import android.text.method.TextKeyListener;
 import android.util.AttributeSet;
 import android.view.KeyEvent;
 import android.view.View;
 import android.view.ViewGroup.MarginLayoutParams;
+import android.view.inputmethod.InputMethodManager;
 
 import com.android.launcher3.DeviceProfile;
 import com.android.launcher3.ExtendedEditText;
@@ -45,6 +48,7 @@ import com.android.launcher3.search.SearchCallback;
 import com.android.launcher3.views.ActivityContext;
 
 import java.util.ArrayList;
+import java.util.Collections;
 
 /**
  * Layout to contain the All-apps search UI.
@@ -53,6 +57,8 @@ public class AppsSearchContainerLayout extends ExtendedEditText
         implements SearchUiManager, SearchCallback<AdapterItem>,
         AllAppsStore.OnUpdateListener, Insettable {
 
+    private boolean mIsSearchSessionActive = false;
+
     private final ActivityContext mLauncher;
     private final AllAppsSearchBarController mSearchBarController;
     private final SpannableStringBuilder mSearchQueryBuilder;
@@ -80,10 +86,37 @@ public class AppsSearchContainerLayout extends ExtendedEditText
         Selection.setSelection(mSearchQueryBuilder, 0);
         setHint(prefixTextWithIcon(getContext(), R.drawable.ic_allapps_search, getHint()));
 
+        addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if (s != null && !s.isEmpty()) {
+                    mIsSearchSessionActive = true;
+                }
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+            }
+        });
+
         mContentOverlap =
                 getResources().getDimensionPixelSize(R.dimen.all_apps_search_bar_content_overlap);
     }
 
+    @Override
+    protected void viewClicked(InputMethodManager imm) {
+        super.viewClicked(imm);
+        if (!mIsSearchSessionActive) {
+            mIsSearchSessionActive = true;
+            // non-null list to trigger animateToSearchState
+            mAppsView.setSearchResults(Collections.emptyList());
+        }
+    }
+
     @Override
     protected void onAttachedToWindow() {
         super.onAttachedToWindow();
@@ -145,6 +178,9 @@ public class AppsSearchContainerLayout extends ExtendedEditText
 
     @Override
     public void resetSearch() {
+        // The doc comment for this method in SearchUiManager says this should close any active
+        // search session.
+        mIsSearchSessionActive = false;
         mSearchBarController.reset();
     }
 
@@ -154,6 +190,11 @@ public class AppsSearchContainerLayout extends ExtendedEditText
         return query.isEmpty();
     }
 
+    @Override
+    public boolean shouldInterceptBackButton() {
+        return mIsSearchSessionActive;
+    }
+
     @Override
     public void preDispatchKeyEvent(KeyEvent event) {
         // Determine if the key event was actual text, if so, focus the search bar and then dispatch
@@ -182,17 +223,40 @@ public class AppsSearchContainerLayout extends ExtendedEditText
 
     @Override
     public void clearSearchResult() {
+        // The doc comment for clearSearchResult in the SearchCallback interface says:
+        // "Called when the search results should be cleared." This is also called when the search
+        // query is empty by AllAppsSearchBarController#afterTextChanged.
+
         // Clear the search query
         mSearchQueryBuilder.clear();
         mSearchQueryBuilder.clearSpans();
         Selection.setSelection(mSearchQueryBuilder, 0);
-        mAppsView.onClearSearchResult();
+
+        // The doc comment for ActivityAllAppsContainerView#onClearSearchResult says, "Invoke when
+        // the current search session is finished," but this is being called in a method called
+        // clearSearchResult. Finishing a search session and clearing the search result should have
+        // different semantics.
+        //
+        // NexusLauncher has similar logic guarding this call in
+        // UniversalSearchInputView#clearSearchResult.
+        if (!mIsSearchSessionActive) {
+            mAppsView.onClearSearchResult();
+        } else {
+            // Must do this or else the latest non-empty search results will remain. This is
+            // normally handled as a result of calling onClearSearchResult
+            mAppsView.setSearchResults(null);
+        }
     }
 
     @Override
     public void setInsets(Rect insets) {
         MarginLayoutParams mlp = (MarginLayoutParams) getLayoutParams();
-        mlp.topMargin = insets.top;
+        DeviceProfile dp = mLauncher.getDeviceProfile();
+        // Use bottom_sheet_handle_area_height instead which is what NexusLauncher's
+        // UniversalSearchInputView does. This puts it closer to the bottom sheet handle, as it
+        // makes it flush against the handle area.
+        mlp.topMargin = getResources().getDimensionPixelSize(
+                R.dimen.bottom_sheet_handle_area_height);
         requestLayout();
     }
 
diff --git a/src/com/android/launcher3/allapps/search/SearchSessionManager.kt b/src/com/android/launcher3/allapps/search/SearchSessionManager.kt
new file mode 100644
index 0000000000..a615650fb8
--- /dev/null
+++ b/src/com/android/launcher3/allapps/search/SearchSessionManager.kt
@@ -0,0 +1,90 @@
+package com.android.launcher3.allapps.search
+
+import android.util.Log
+import android.view.WindowInsets
+import com.android.launcher3.views.ActivityContext
+import com.android.launcher3.AbstractFloatingView
+
+/**
+ * Manages search session for all apps drawer.
+ *
+ * If we need to keep track of state, turn this into a class and add to various Dagger modules.
+ * NexusLauncher does something more complicated, since they want to hold reference to the
+ * ActivityContext
+ */
+object SearchSessionManager {
+    private const val TAG = "LSearchSessionManager"
+
+    private inline fun verboseLog(makeLog: () -> String) {
+        if (Log.isLoggable(TAG, Log.VERBOSE)) {
+            Log.v(TAG, makeLog())
+        }
+    }
+
+    /**
+     * Returns true if the search session handled the back invocation. If [doActions] is false,
+     * then this method just becomes a query on whether the back invocation would be handled.
+     *
+     * [doActions] will be false on calls for things such as predictive back gesture queries.
+     */
+    @JvmStatic
+    fun handleAllAppsSearchBackInvoked(context: ActivityContext, doActions: Boolean): Boolean {
+        verboseLog { "isAllAppsBackButtonIntercepted(ctx, $doActions)" }
+
+        val appsView = context.appsView ?: return false
+        if (!appsView.isInAllApps) {
+            return false
+        }
+
+        // This branch is present in NexusLauncher, but doesn't seem to be triggered right now in
+        // AOSP Launcher3.
+        val noInterceptTopOpenView = AbstractFloatingView.getTopOpenViewWithType(
+            context,
+            AbstractFloatingView.TYPE_TOUCH_CONTROLLER_NO_INTERCEPT
+        )
+        if (noInterceptTopOpenView != null) {
+            Log.d(TAG, "topNoInterceptOpenView ${noInterceptTopOpenView::class.java.name}")
+            if (noInterceptTopOpenView.canHandleBack()) {
+                Log.d(TAG, "topNoInterceptOpenView canHandleBack")
+                if (doActions) {
+                    noInterceptTopOpenView.onBackInvoked()
+                }
+                return true;
+            }
+        }
+
+        val manager = appsView.searchUiManager
+        val editText = manager.editText ?: return false
+        val isImeVisible = editText.rootWindowInsets?.isVisible(WindowInsets.Type.ime())
+            ?: return false
+        if (!manager.shouldInterceptBackButton()) {
+            verboseLog { "manager not allowing intercepting back, isImeVisible $isImeVisible" }
+            if (isImeVisible) {
+                if (doActions) {
+                    editText.hideKeyboard()
+                }
+                // handled back invoked since we hid the keyboard
+                return true
+            }
+            return false
+        }
+
+        // appsView.requestFocus call is made here in NexusLauncher
+        if (doActions) {
+            appsView.requestFocus()
+        }
+
+        verboseLog { "handling back button, isImeVisible $isImeVisible" }
+        // condition used in NexusLauncher
+        if (editText.text.isEmpty() || !isImeVisible) {
+            if (doActions) {
+                Log.d(TAG, "ending search session from back button");
+                manager.resetSearch()
+            }
+        } else if (doActions) {
+            editText.hideKeyboard()
+        }
+
+        return true
+    }
+}
diff --git a/src/com/android/launcher3/deviceprofile/WorkspaceProfileNonResponsiveFactory.kt b/src/com/android/launcher3/deviceprofile/WorkspaceProfileNonResponsiveFactory.kt
index fcd789e30b..4dad53afa7 100644
--- a/src/com/android/launcher3/deviceprofile/WorkspaceProfileNonResponsiveFactory.kt
+++ b/src/com/android/launcher3/deviceprofile/WorkspaceProfileNonResponsiveFactory.kt
@@ -50,6 +50,11 @@ import kotlin.math.min
  */
 object WorkspaceProfileNonResponsiveFactory {
 
+    private fun shouldHideWorkspaceItemsLabel(deviceProperties: DeviceProperties): Boolean {
+        return deviceProperties.deviceConfiguration.isWorkspaceItemsLabelHidden &&
+                com.android.systemui.shared.Flags.workspaceItemsLabelHidden()
+    }
+
     fun createWorkspacePadding(
         isVerticalLayout: Boolean,
         isSeascape: Boolean,
@@ -166,6 +171,7 @@ object WorkspaceProfileNonResponsiveFactory {
         panelCount: Int,
         scale: Float,
     ): WorkspaceProfile {
+        val hideWorkspaceItemsLabel = shouldHideWorkspaceItemsLabel(deviceProperties)
         val cellLayoutBorderSpacePx = Point(0, 0)
         var cellSize =
             calculateCellSize(
@@ -182,15 +188,25 @@ object WorkspaceProfileNonResponsiveFactory {
                 isVerticalLayout -> 0
                 else -> res.getDimensionPixelSize(R.dimen.dynamic_grid_left_right_margin)
             }
-        var iconDrawablePaddingPx =
-            (getNormalizedIconDrawablePadding(iconSizePx, iconDrawablePaddingOriginalPx) *
-                    iconScale)
-                .toInt()
+        var iconDrawablePaddingPx = when {
+            hideWorkspaceItemsLabel -> 0
+            else -> {
+                (getNormalizedIconDrawablePadding(iconSizePx, iconDrawablePaddingOriginalPx) *
+                        iconScale)
+                    .toInt()
+            }
+        }
+
+        val iconTextSize = when {
+            hideWorkspaceItemsLabel -> 0
+            else -> iconTextSizePx
+        }
+
         val cellWidthPx = iconSizePx + iconDrawablePaddingPx
         var cellHeightPx =
             (getIconSizeWithOverlap(iconSizePx) +
                 iconDrawablePaddingPx +
-                Utilities.calculateTextHeight(iconTextSizePx.toFloat()))
+                Utilities.calculateTextHeight(iconTextSize.toFloat()))
         val cellPaddingY: Int = (cellSize.y - cellHeightPx) / 2
         if (
             iconDrawablePaddingPx > cellPaddingY &&
@@ -259,7 +275,7 @@ object WorkspaceProfileNonResponsiveFactory {
             // Workspace icons
             iconScale = iconScale,
             iconSizePx = iconSizePx,
-            iconTextSizePx = iconTextSizePx,
+            iconTextSizePx = iconTextSize,
             iconDrawablePaddingPx = iconDrawablePaddingPx,
             cellScaleToFit = cellScaleToFit,
             cellWidthPx = cellWidthPx,
@@ -267,7 +283,7 @@ object WorkspaceProfileNonResponsiveFactory {
             cellLayoutBorderSpacePx = cellLayoutBorderSpacePx,
             desiredWorkspaceHorizontalMarginPx = desiredWorkspaceHorizontalMarginPx,
             cellYPaddingPx = -1,
-            maxIconTextLineCount = 1,
+            maxIconTextLineCount = if (hideWorkspaceItemsLabel) 0 else 1,
             iconCenterVertically = false,
             gridVisualizationPaddingX =
                 res.getDimensionPixelSize(R.dimen.grid_visualization_horizontal_cell_spacing),
@@ -303,6 +319,7 @@ object WorkspaceProfileNonResponsiveFactory {
             panelCount = panelCount,
             cellSize = cellSize,
             scale = scale,
+            isItemsLabelHidden = deviceProperties.deviceConfiguration.isWorkspaceItemsLabelHidden,
             flingToDeleteThresholdVelocity =
                 res.getDimensionPixelSize(R.dimen.drag_flingToDeleteMinVelocity),
         )
@@ -326,6 +343,7 @@ object WorkspaceProfileNonResponsiveFactory {
         isSeascape: Boolean,
         hotseatProfile: HotseatProfileInitialValues,
     ): WorkspaceProfile {
+        val hideWorkspaceItemsLabel = shouldHideWorkspaceItemsLabel(deviceProperties)
         val cellLayoutBorderSpacePx =
             Point(
                 pxFromDp(inv.borderSpaces[typeIndex].x, metrics, scale),
@@ -337,12 +355,21 @@ object WorkspaceProfileNonResponsiveFactory {
                 isVerticalLayout -> 0
                 else -> pxFromDp(inv.horizontalMargin[typeIndex], metrics)
             }
-        var iconTextSizePx = iconTextSizePxParam
+        var iconTextSizePx = when {
+            hideWorkspaceItemsLabel -> 0
+            else -> iconTextSizePxParam
+        }
+
         var iconSizePx = iconSizePxParam
-        var iconDrawablePaddingPx =
-            (getNormalizedIconDrawablePadding(iconSizePx, iconDrawablePaddingOriginalPx) *
-                    iconScale)
-                .toInt()
+        var iconDrawablePaddingPx = when {
+            hideWorkspaceItemsLabel -> 0
+            else -> {
+                (getNormalizedIconDrawablePadding(iconSizePx, iconDrawablePaddingOriginalPx) *
+                        iconScale)
+                    .toInt()
+            }
+        }
+
         var cellWidthPx = pxFromDp(inv.minCellSize.get(typeIndex).x, metrics, scale)
         var cellHeightPx = pxFromDp(inv.minCellSize.get(typeIndex).y, metrics, scale)
 
@@ -463,7 +490,7 @@ object WorkspaceProfileNonResponsiveFactory {
             cellLayoutBorderSpacePx = cellLayoutBorderSpacePx,
             desiredWorkspaceHorizontalMarginPx = desiredWorkspaceHorizontalMarginPx,
             cellYPaddingPx = max(0, (cellHeightPx - cellContentHeight)) / 2,
-            maxIconTextLineCount = 1,
+            maxIconTextLineCount = if (hideWorkspaceItemsLabel) 0 else 1,
             iconCenterVertically = isVerticalLayout,
             gridVisualizationPaddingX =
                 res.getDimensionPixelSize(R.dimen.grid_visualization_horizontal_cell_spacing),
@@ -499,6 +526,7 @@ object WorkspaceProfileNonResponsiveFactory {
             panelCount = panelCount,
             cellSize = cellSize,
             scale = scale,
+            isItemsLabelHidden = deviceProperties.deviceConfiguration.isWorkspaceItemsLabelHidden,
             flingToDeleteThresholdVelocity =
                 res.getDimensionPixelSize(R.dimen.drag_flingToDeleteMinVelocity),
         )
diff --git a/src/com/android/launcher3/util/KeyboardShortcutsDelegate.java b/src/com/android/launcher3/util/KeyboardShortcutsDelegate.java
index 9c18144c96..af2e3349fd 100644
--- a/src/com/android/launcher3/util/KeyboardShortcutsDelegate.java
+++ b/src/com/android/launcher3/util/KeyboardShortcutsDelegate.java
@@ -32,6 +32,7 @@ import com.android.launcher3.LauncherState;
 import com.android.launcher3.R;
 import com.android.launcher3.Utilities;
 import com.android.launcher3.accessibility.BaseAccessibilityDelegate;
+import com.android.launcher3.allapps.search.SearchSessionManager;
 import com.android.launcher3.popup.WorkspaceLongPressOptions;
 import com.android.launcher3.testing.shared.TestProtocol;
 
@@ -119,8 +120,11 @@ public class KeyboardShortcutsDelegate {
                 topView.close(/* animate= */ true);
                 return true;
             } else if (mLauncher.getAppsView().isInAllApps()) {
-                // Close all apps if there are no open floating views.
-                mLauncher.getStateManager().goToState(NORMAL, true);
+                // Close all apps if there are no open floating views. If search is active, close
+                // that first.
+                if (!SearchSessionManager.handleAllAppsSearchBackInvoked(mLauncher, true)) {
+                    mLauncher.getStateManager().goToState(NORMAL, true);
+                }
                 return true;
             } else if (mLauncher.isInState(LauncherState.OVERVIEW)
                     || mLauncher.isInState(LauncherState.OVERVIEW_SPLIT_SELECT)) {
